enterprise/server/auth/auth.go

```diff
@@ -252,17 +259,84 @@ func (a *authenticator) verifyTokenAndExtractUser(ctx context.Context, jwt strin
 	return extractToken(a.issuer, validToken)
 }
 
+type apiKeyGroupCacheEntry struct {
+	data         interfaces.APIKeyGroup
+	expiresAfter time.Time
+}
+
+// apiKeyGroupCache is a cache for API Key -> Group lookups. A single Bazel invocation
+// can generate large bursts of RPCs, each of which needs to be authed.
+// There's no need to go to the database for every single request as this data
+// rarely changes.
+type apiKeyGroupCache struct {
+	// Note that even though we base this off an LRU cache, every entry has a hard expiration
+	// time to force a refresh of the underlying data.
+	lru *lru.LRU
+	ttl time.Duration
+	mu  sync.RWMutex
+}
+
+func newAPIKeyGroupCache(configurator *config.Configurator) (*apiKeyGroupCache, error) {
+	var err error
+	var ttl time.Duration
+	if configurator.GetAuthAPIKeyGroupCacheTTL() != "" {
+		ttl, err = time.ParseDuration(configurator.GetAuthAPIKeyGroupCacheTTL())
+		if err != nil {
+			return nil, fmt.Errorf("invalid API Key -> Group cache TTL [%s]: %v", configurator.GetAuthAPIKeyGroupCacheTTL(), err)
+		}
+	} else {
+		ttl = defaultAPIKeyGroupCacheTTL
+	}
+
+	sizeFunc := func(k, v interface{}) int64 { return 1 }
+	lru, err := lru.NewLRU(apiKeyGroupCacheSize, nil /* onEvict */, nil /* onAdd */, sizeFunc)
+	if err != nil {
+		return nil, fmt.Errorf("error initializing API Key -> Group cache: %v", err)
+	}
+	return &apiKeyGroupCache{lru: lru, ttl: ttl}, nil
+}
+
+func (c *apiKeyGroupCache) Get(apiKey string) (akg interfaces.APIKeyGroup, ok bool) {
+	c.mu.RLock()
+	v, ok := c.lru.Get(apiKey)
+	c.mu.RUnlock()
+	if !ok {
+		return nil, ok
+	}
+	entry, ok := v.(*apiKeyGroupCacheEntry)
+	if !ok {
+		// Should never happen.
+		log.Printf("Data in cache was of wrong type, got type %T", v)
+		return nil, false
+	}
+	if time.Now().After(entry.expiresAfter) {
+		return nil, false
+	}
+	return entry.data, true
+}
+
+func (c *apiKeyGroupCache) Add(apiKey string, apiKeyGroup interfaces.APIKeyGroup) {
+	c.mu.Lock()
+	c.lru.Add(apiKey, &apiKeyGroupCacheEntry{data: apiKeyGroup, expiresAfter: time.Now().Add(c.ttl)})
+	c.mu.Unlock()
+}
+
 type OpenIDAuthenticator struct {
-	env            environment.Env
-	myURL          *url.URL
-	authenticators []*authenticator
+	env              environment.Env
+	myURL            *url.URL
+	authenticators   []*authenticator
+	apiKeyGroupCache *apiKeyGroupCache
```

Comment:
What do you think about putting this (apiKeyGroupCache) in the authDB instead of in this auth code?

Reply:
It's a good suggestion but it's slightly annoying in practice because AuthDB is in the internal repo.

We ought to just move AuthDB to the OSS repo to make things simple but don't think it's worth the effort for this change.

Reply:
OK, SGTM for now. Easy to move later!